### ISSUE TITLE
runtime: End to end tests should run against both the native and the wasm runtimes

### ISSUE THREAD

**Issue by NunoAlexandre** [radicle-dev/radicle-registry#440]
Part of #434 

Have the end to end tests running against both the native and the wasm runtimes to verify that both are supported. Doing so should verify that the desired behaviour for the ledger is met when running both the latest native version with new changes and an older version running through WASM, that new versions are backwards-compatible, etc.

#### How?
A way of achieving this is to have a script that starts the dev node using the native runtime to produce blocks (#438), runs the end to end tests, stop the node, reset the dev chain (?), start the node again now using the wasm runtime to produce blocks and run the tests again. Both runs should succeed.

#### What happens when they differ?

Following the proposed above would make comparing failed tests quite difficult. Say than running a specific end to end test against the native runtime produces foo and against the wasm runtime, it produces bar. Since they don't run in parallel, it will be difficult to process failed tests. Something to keep in mind.

**Comment by CodeSandwich** [comment]
I remember discussing whether we should use the native runtime at all. The complexity and risk are high, but AFAIK the only benefit is improved performance. We don't have smart contracts, our functionality is very light on calculations.

**Comment by NunoAlexandre** [comment]
I like that consideration. However, the idea of doing this and using the native was a matter of performance but a matter of checking whether the native (running the very last revision of a feature branch) behaved and produced the same results and the outdated wasm.

**Comment by geigerzaehler** [comment]
I agree with @CodeSandwich that there is no need to run native code for performance. And I agree with @NunoAlexandre that we should verify that the two versions match. See also my [comment on #441](https://github.com/radicle-dev/radicle-registry/issues/441#issuecomment-627219893)

However this only applies if the changes are just implementation changes. If not we expect the native runtime to differ from an old Wasm runtime. In that case we will need to update the Wasm, too.

**Comment by geigerzaehler** [comment]
I’ve tried to implement this an I noticed that even with `ExecutionStrategy::Both` for block production the node does not warn of a difference. (To simulate a difference I removed actually updating the state when registering a user.) Need to investigate this further.

**Comment by geigerzaehler** [comment]
While working on #488 I could see that the node was detecting a difference in the runtime versions and emitting a warning.

**Comment by CodeSandwich** [comment]
I was trying to figure out what's the goal of cross-testing of native and WASM runtimes on the CI:

- Testing if the compiler produces equivalent WASM and native binaries from the same code. In my opinion this is overly paranoid and shouldn't be our goal.
- Testing if the code was changed and we need to somehow publish these changes. This seems to be a valuable goal.

There can be two kinds of runtime changes:
- Requiring bumping `impl_version`
- Requiring bumping `spec_version`

`impl_version` should be bumped whenever there's any change in runtime at all, either in the code or in the dependencies. This should be detectable by something as simple as a script checking in git if any file in `/runtime` was modified after the last modification of `/runtime/latest.wasm`. Because git's granularity of time is limited to whole commits (file modification time isn't stored), regeneration of `latest.wasm` should be a separate commit, preferably the last one in a merged branch. This can be easily cheated, but should suffice as a genuine mistake detector.

`spec_version` should be bumped when the above applies, but additionally there are changes in the runtime's observable behavior. Our e2e tests can't guarantee catching all the subtle changes and shouldn't be the final argument when deciding NOT to bump `spec_version`, this still needs to be a developer's decision.

If there's a change to the runtime, we need to test the resulting new client against the runtime from the last release. This will guarantee backward compatibility with a single previous version of the runtime, which was previously released. Checking just the previous one is enough, because release of a runtime with a new `spec_version` needs to be accompanied by publishid a new CLI. After all the client is based on the native runtime and can't be updated with a transaction like the node.

In this setup there's no need to test against a node running a native runtime, because we can easily guarantee that the WASM version is up to date. Additionally we need a script, which will fetch `latest.wasm` from the last release and use it for the second run to prove backward compatibility of the client library.

In conclusion the CI needs:
- `latest.wasm` being updated test
- regular workspace test with the WASM runtime
- regular workspace test with the WASM runtime from the previous release

This comment's scope slightly overlaps https://github.com/radicle-dev/radicle-registry/issues/489.

**Comment by geigerzaehler** [comment]
The initial goal for this issue was to just run the tests against native runtime and the wasm runtime to ensure compatibility between to different implementation versions. Since we’re no re-evaluating that goal within the broader scope of testing runtime compatibility [on radicle.community](https://radicle.community/t/runtime-updates-workflow-and-testing/224) I’m closing this issue.
